trlx/orchestrator/ppo_orchestrator.py

```diff
@@ -103,54 +114,98 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):
+                            all_tokens,
+                            attention_mask=attention_mask,
+                            position_ids=position_ids,
+                            return_dict=False,
+                        )
+                    else:
+                        ref_logits, _, *_ = self.ref_model(
+                            all_tokens,
+                            attention_mask=attention_mask,
+                            position_ids=position_ids,
+                            return_dict=False,
+                        )
+                        ref_logits = ref_logits.to(self.trainer.accelerator.device)
+
+            if self.trainer.config.model.model_arch_type == "encoder_decoder":
+                logprobs = logprobs_from_logits(logits, response_tensors)
+                ref_logprobs = logprobs_from_logits(
+                    ref_logits, response_tensors
+                )
+                values = values.cpu()
+            else:
+            
+                logprobs = logprobs_from_logits(logits[:, :-1, :], all_tokens[:, 1:])
+                ref_logprobs = logprobs_from_logits(
+                    ref_logits[:, :-1, :], all_tokens[:, 1:]
+                )
+                values = values.cpu()[:, :-1]
             n = samples.shape[0]
-            values = values.cpu()[:, :-1]
             logprobs = logprobs.cpu()
             ref_logprobs = ref_logprobs.cpu()
             query_tensors = query_tensors.cpu()
             response_tensors = response_tensors.cpu()
 
-            start = query_tensors.shape[1] - 1
-            ends = start + attention_mask[:, start:].sum(1)
-            all_values = [values[ix, start : ends[ix]] for ix in range(n)]
-            all_logprobs = [logprobs[ix, start : ends[ix]] for ix in range(n)]
+            if self.trainer.config.model.model_arch_type == "encoder_decoder":
+                all_values = values
+                all_logprobs = logprobs
+            else:
+                start = query_tensors.shape[1] - 1
+                ends = start + attention_mask[:, start:].sum(1)
+                all_values = [values[ix, start : ends[ix]] for ix in range(n)]
+                all_logprobs = [logprobs[ix, start : ends[ix]] for ix in range(n)]
 
             # Compute rewards
             rewards = -self.trainer.kl_ctl.value * (logprobs - ref_logprobs)
             all_rewards = [None] * n
-            for ix in range(n):
-                rs = rewards[ix][start : ends[ix]]
-                rs[-1] = scores[ix]
-                all_rewards[ix] = rs
-
+            if self.trainer.config.model.model_arch_type == "encoder_decoder":
+                for ix in range(n):
+                    rs = rewards[ix]
+                    rs[-1] = scores[ix]
+                    all_rewards[ix] = rs
+            else:
+                for ix in range(n):
+                    rs = rewards[ix][start : ends[ix]]
+                    if len(rs) == 0: # fix for empty responses
```

Comment:
Ah, nice catch. CC @reciprocated : this was the issue I came across a while back on cases where `start == ends[idx]`; were just empty responses.